wit/world.wit

```diff
@@ -1,11 +1,53 @@
-package wasi:cloud-core;
-
-world cloud-core {
-    include wasi:keyvalue;
-    include wasi:blob-store;
-    include wasi:messaging;
-    include wasi:sql;
-    include wasi:runtime-config;
-    include wasi:distributed-locking;
-    include wasi:http;
+package wasi:cloud-core@0.1.0;
```

Comment:
I'm not sure I follow the reasoning behind this change. As far as I understood, this world was meant to be the set of all things a host/runtime could implement to support `wasi-cloud`. This seems to move away from it and gives things names that don't quite make sense. As a user or implementor, what if I want to just produce messages and access a keyvalue store? Where does that fit? What if I just need a sql database and I run on a timer (not triggered by an inbound message or http request). It seems really odd as well that as a user, if I want to use a blobstore interface, why would I be looking for a world called `http`?

Can you walk me through the reasoning for this change? I would love to be wrong here, but as it stands now I don't think things really make sense

Reply:
IIUC, there must be a exported function for the guest component to implement targeting a world. For example, the `wasi:cli/command` world provides a `run` function and the `wasi:http/proxy` world provides a `incoming-handler` exported function. 

I was thinking that the `wasi:cloud-core` world should follow the same pattern as if the component is either handling HTTP requests or handling messaging events, or something else (maybe I should add an exported `run` function?). So that's basically the reasonsing that I made this change. 

Before this change, the `cloud-core` world would have both `wasi:http/incoming-handler` and `wasi:messaging/handler` exported functions, which in this case, the component has to implement both. Correct me if I am wrong here.

Reply:
> As a user or implementor, what if I want to just produce messages and access a keyvalue store?

As an user, if you just want to produce messages and access a keyvalue store, you will just need to use the wasi:messaging and wasi:keyvalue APIs provided by wasi:cloud-core. But the point of this change is that your component must have something like an entrypoint for the host to call, think about a `main()` function. 

As an implementor, you will have to provide the implementation of all the imports from this world, including wasi:messaging and wasi:keyvalue.

Reply:
> It seems really odd as well that as a user, if I want to use a blobstore interface, why would I be looking for a world called http?

Are you concerned primarily on the naming of the world? I get it. How about I revert back the world name to `cloud-core`, but remove the exported functions? 

It would look like

```wit
world cloud-core {
    /// The capability to interact with key-value stores
    include wasi:keyvalue/imports;

    /// The capability to interact with blob stores
    include wasi:blobstore/imports;

    /// The capability to send messages to message broker proiders
    import wasi:messaging/producer;

    /// The capability to receive messages from message broker providers
    import wasi:messaging/consumer;

    /// The capability to interact with SQL databases
    include wasi:sql/imports;

    /// The capability to interact with runtime configuration providers
    include wasi:runtime-config/imports@0.1.0;

    /// The capability to send/receive HTTP requests
   import wasi:http/outgoing-handler@0.2.0-rc-2023-12-05;
}
```

Reply:
Right, `include` pulls those in transitively, which I always forget that. However, there are a whole bunch of different ways people could be running applications other than triggered by an HTTP request or a message. Two that come to mind are:

- Custom interfaces. I could definite a custom interface that I use to trigger. This would definitely be used by something doing a plugin model, I also know it would happen in wasmCloud. But the point is anyone can define an arbitrary interface
- cli/command (which you mentioned). People could want to run this component from the command line

It is also important to note that people can bring in the individual worlds they want here, but for the main wasi-core, we don't want to _require_ that you have them exposed via http or messaging. That might be the case for a FaaS, but not for all types of cloud applications.

Reply:
Agreed. Do you think that my proposed `cloud-core` world, which has a bunch of `import`s, but no `export`s, could be embedded to custom worlds that all sorts of applications need?

Reply:
This is maybe a heretical thought, but it might be better to not have a top level world, especially since it is likely that people will pick and choose which of the wasi-cloud-core interfaces they want to use.

That would mean this repo would just be a "meta" repo to discuss which interfaces fall under this umbrella and maybe even contain some of the common worlds in the future. But I think I'd prefer not defining worlds like this until we have more implementations and real world use cases under our belt

Reply:
> But I think I'd prefer not defining worlds like this until we have more implementations and real world use cases under our belt

Allow me to disagree here, I believe that having a world in this repo is pretty important, and here is why. Defining a spec (a WIT world) in wasi-cloud-core repo allows both the host implementation and guest components to align to a single target, which is a key principle of portability that wasi-cloud adheres to. A component targets wasi-cloud world should be expected to run on any hosts that implement wasi-cloud worlds and this is the portability story we should be telling. If there are no worlds defined in this repo, no contract between the guest and the host, then we don't have a good story for porting components from one runtime to another. 

The WIT files can be changed, yes, but they should be versioned so they should provide a certain degree of stability.

Reply:
After talking some more with a bunch of people at the plumbers summit, I do think we need a world, but I still stick by the argument that they way the worlds are designed here are confusing for people to implement. @lukewagner I selfishly would love to get your thoughts in written form about the difference between imports and exports in worlds, but perhaps that is a discussion that is better done elsewhere

Reply:
You may regret asking, but here's the thoughts (and happy to delete this comment and copy/aste it to a new issue if you want):

So as some groundwork: I expect we agree that it's useful to factor out a `wasi:cloud-core/imports` world that imports all the common cloud-core stuff.  Then separately we can define one or more worlds that describe exports and `include imports;`.

But the more nuanced question that comes up in the context of cloud-core is: when a `world` exports two interfaces, do we say that a component targeting the world is (1) required to export both or (2) allowed-but-not-required to export both.

Intuitively, if we imagine a cloud-core host that supports triggers (export calls) for HTTP, keyvalue-store watch notifications, WebSockets, MQTT, gRPC, etc, we want our world to somehow express the total set of supported triggers without forcing every component to define an export for every single one.  Or, from another angle, when a host adds a new feature that adds a new trigger, we want to *allow* new components to take advantage of it (defining a new export), but obviously we don't want to declare all existing components that don't export the just-added trigger invalid (this is a sort of backwards-compatibility requirement).  Also, iirc, Wasmtime currently doesn't require all exports of a world to be exported by a component, and instead makes it a dynamic property that the embedder has to test before calling.  Based on all this, we might be inclined to choose (2).

However, I think we actually need (1) and we can address the use cases above by adding the ability to make imports *and* exports explicitly `optional` (in WIT and in the underlying C-M and runtime).

The reason I think we need (1) is: imagine a component C that imports and instantiates another unknown component X.  In [WAC](https://github.com/peterhuene/wac), you'd express importing an unknown component via `import X: some-world;` where `some-world` refers to a `world`, and then C could instantiate `X` with `let x = new X { ... }`.  Now, if `some-world` exports a function `foo`, C will expect `X` to *definitely* export `foo` (so C can pass `foo` to another child component or re-export `foo` from C).  But if world exports are always optional (as in (2)), then seemingly-valid values of `X` (that target `some-world` and don't export `foo`) will cause unexpected errors.  Thus, world exports are mandatory (which, once you know a `world` is just a component type, and component types are fancy function types, follows from the classic [function subtyping rules](https://en.wikipedia.org/wiki/Subtyping#Function_types)).

Instead, I think we need to add optionality explicitly via a new `optional` prefix to types.  E.g., I'm imagining that you could write:
```wit
world some-world {
  import wasi:http/outgoing-handler;
  import wasi:keyvalue/eventual;
  import optional wasi:webgpu/device;
  export optional wasi:http/incoming-handler;
  export optional wasi:websockets/incoming-handler;
}
```
which would allow a component targeting `some-world` to export zero, one or both of the HTTP and WebSockets hander interfaces.  And now if our component C imports an unknown component X of type `some-world`, C will know that it can't assume either export exists.  (The Canonical ABI of `optional` would export a dynamic "did I get it?" core function that can be used to probe dynamically.)

A related question comes up if I include two worlds with disjoint exports, e.g.:
```wit
world w1 { export foo: func() }
world w2 { export bar: func() }
world w3 { include w1; include w2; }
```
the expanded final type of `w3` would naturally be:
```wit
world w3 {
  export foo: optional foo()
  export bar: optional bar()
}
```

So finally returning to the concrete question "what should wasi-cloud-core do in the short term", I'd suggest:
* We define `wasi:cloud-core/imports` as suggested above.
* We start working to add `option` to WIT and the C-M, motivated by the above.
* *In the meantime/short-term*, we define a handful of worlds that export meaningful subsets of { `wasi:http/ingcoming-handler`, `wasi:keyvalue/handle-watch`, `wasi:messaging/messaging-guest`, ... }.
  * For example, `wasi:cloud-core/http-service` could export `wasi:http/incoming-handler`.
  * But also, there should be one world that exports them all, say, `wasi:cloud-core/service`.
* When doing host bindgen, you'd pass `wasi:cloud-core/service`, and rely on the "all exports are treated as optional" behavior mentioned above so the host could run components exporting any subset.

Bullet 3 is a brute force workaround to the absence of bullet 2 and obviously doesn't scale as we add new exportable interfaces (the size of powersets grows exponentially :).  But maybe it's good enough for a while and also helps us properly motivate adding `option`.

Reply:
Updated the PR to have
1. `wasi:cloud-core/imports` world
2. `wasi:cloud-core/http-service` world
3. and `wasi:cloud-core/messaging-service` world

Reply:
Cool; that all looks good to me.

Reply:
Could you please take another look? @thomastaylor312

Reply:
@lukewagner Really love these thoughts. We should definitely copy/link them in an issue for WIT (or open one if it doesn't exist).